src/main/java/core/basesyntax/Car.java

```diff
@@ -1,13 +1,81 @@
 package core.basesyntax;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Objects;
 
 /**
  * Make this class immutable. See requirements in the README file
  */
-public class Car {
-    private int year;
-    private String color;
-    private List<Wheel> wheels;
-    private Engine engine;
+public final class Car {
+    private final int year;
+    private final String color;
+    private final List<Wheel> wheels;
+    private final Engine engine;
+
+    public Car(int year, String color, List<Wheel> wheels, Engine engine) {
+        this.year = year;
+        this.color = color;
+        this.wheels = copyOfList(wheels);
+        this.engine = engine == null ? null : engine.clone();
+    }
+
+    public String getColor() {
+        return color;
+    }
+
+    public Engine getEngine() {
+        return engine == null ? null : engine.clone();
+    }
+
+    public Car changeEngine(Engine engine) {
+        return new Car(year, color, wheels, engine);
+    }
+
+    public List<Wheel> getWheels() {
+        return copyOfList(wheels);
+    }
+
+    public Car changeColor(String anotherColor) {
+        return new Car(year, anotherColor, wheels, engine);
+    }
+
+    public int getYear() {
+        return year;
+    }
+
+    public Car addWheel(Wheel wheel) {
+        List<Wheel> otherWheels = getWheels();
```

Comment:
```suggestion
        List<Wheel> newWheel = getWheels(); 
```
Bad naming for list and reduntant method call.